server/src/main/scala/com/cloudera/livy/sessions/Session.scala

```diff
@@ -47,26 +47,36 @@ abstract class Session(val id: Int, val owner: String, val livyConf: LivyConf) e
 
   protected implicit val executionContext = ExecutionContext.global
 
+  protected var _appId: Option[String] = None
+
   private var _lastActivity = System.nanoTime()
 
   // Directory where the session's staging files are created. The directory is only accessible
   // to the session's effective user.
   private var stagingDir: Path = null
 
+  def appId: Option[String] = _appId
+
   def lastActivity: Long = state match {
     case SessionState.Error(time) => time
     case SessionState.Dead(time) => time
     case SessionState.Success(time) => time
     case _ => _lastActivity
   }
 
-  val timeout: Long = TimeUnit.HOURS.toNanos(1)
+  def logLines(): IndexedSeq[String]
+
+  def recordActivity(): Unit = {
+    _lastActivity = System.nanoTime()
+  }
 
   def state: SessionState
 
   def stop(): Future[Unit] = Future {
     try {
+      info(s"Stopping $this...")
```

Comment:
When something goes wrong in production, how do we debug Livy without logging user operations?